### PR TITLE
[NOD-1319] Reimplement kaspad's RPC in gRPC

### PR DESCRIPTION
**Pull request by stasatdaglabs**

### CODE REVIEW

File: infrastructure/network/rpcclient/rpc_send_raw_transaction.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"bytes"
+	"encoding/hex"
+	"github.com/kaspanet/kaspad/app/appmessage"
+)
+
+// SendRawTransaction sends an RPC request respective to the function's name and returns the RPC server's response
+func (c *RPCClient) SendRawTransaction(msgTx *appmessage.MsgTx) (*appmessage.SendRawTransactionResponseMessage, error) {
+	transactionHex := ""
+	if msgTx != nil {
```
Comment (someone235) [inline infrastructure/network/rpcclient/rpc_send_raw_transaction.go:12]: Why do you allow sending nil transactions

File: infrastructure/network/rpcclient/rpc_submit_block.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import (
+	"encoding/hex"
+	"github.com/kaspanet/kaspad/app/appmessage"
+	"github.com/kaspanet/kaspad/util"
+)
+
+// SubmitBlock sends an RPC request respective to the function's name and returns the RPC server's response
+func (c *RPCClient) SubmitBlock(block *util.Block) error {
+	blockHex := ""
+	if block != nil {
```
Comment (someone235) [inline infrastructure/network/rpcclient/rpc_submit_block.go:12]: Why do you allow sending nil blocks?

File: infrastructure/network/rpcclient/log.go [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package rpcclient
+
+import (
+	"github.com/kaspanet/kaspad/infrastructure/logger"
+	"github.com/kaspanet/kaspad/util/panics"
+)
+
+var log, _ = logger.Get(logger.SubsystemTags.RPCC)
```
Comment (someone235) [inline infrastructure/network/rpcclient/log.go:8]: You need to define UseLogger function, otherwise you'll be able to use the RPCClient only with kaspad logger, which is not very useful

File: cmd/kaspaminer/log.go [hunk 4]
```diff
@@ -26,7 +25,3 @@ func initLog(logFile, errLogFile string) {
 		os.Exit(1)
 	}
 }
-
-func enableRPCLogging() {
```
Comment (someone235) [inline cmd/kaspaminer/log.go:30]: Why did you delete it?

File: testing/integration/64_incoming_connections_test.go [hunk 5]
```diff
@@ -36,19 +36,19 @@ func Test64IncomingConnections(t *testing.T) {
...
 			if blockAdded {
 				t.Fatalf("Single bully reported block added twice")
 			}
 			blockAdded = true
 			blockAddedWG.Done()
 		}
+
+		err := bully.rpcClient.RegisterForBlockAddedNotifications(onBlockAdded)
+		if err != nil {
+			t.Fatalf("Error from NotifyBlocks: %+v", err)
```
Comment (someone235) [inline testing/integration/64_incoming_connections_test.go:50]: Change NotifyBlocks->RegisterForBlockAddedNotifications

File: testing/integration/notifications_test.go [hunk 6]
```diff
@@ -6,10 +6,9 @@ import (
 	"github.com/kaspanet/kaspad/app/appmessage"
 )
 
-func setOnBlockAddedHandler(t *testing.T, harness *appHarness, handler func(header *appmessage.BlockHeader)) {
-	err := harness.rpcClient.NotifyBlocks()
+func setOnBlockAddedHandler(t *testing.T, harness *appHarness, handler func(notification *appmessage.BlockAddedNotificationMessage)) {
+	err := harness.rpcClient.RegisterForBlockAddedNotifications(handler)
 	if err != nil {
 		t.Fatalf("Error from NotifyBlocks: %s", err)
```
Comment (someone235) [inline testing/integration/notifications_test.go:12]: NotifyBlocks->RegisterForBlockAddedNotifications

File: testing/integration/rpc_test.go [hunk 7]
```diff
@@ -1,37 +1,24 @@
...
-	rpcclient "github.com/kaspanet/kaspad/infrastructure/network/rpc/client"
+	"github.com/kaspanet/kaspad/infrastructure/network/rpcclient"
+	"time"
 )
 
-type rpcClient struct {
-	*rpcclient.Client
-	onBlockAdded func(*appmessage.BlockHeader)
-}
+const testTimeout = 1 * time.Second
```
Comment (someone235) [inline testing/integration/rpc_test.go:8]: This variable is in the package scope, so please change it to rpcTimeout or something else more meaningful

**Review by someone235 — CHANGES_REQUESTED** [review]

File: app/appmessage/rpc_notify_chain_changed.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+// ChainChangedNotificationMessage is an appmessage corresponding to
+// its respective RPC message
+type ChainChangedNotificationMessage struct {
+	baseMessage
+	RemovedChainBlockHashes []string
+	AddedChainBlocks        []*ChainChangedChainBlock
+}
+
+// ChainChangedChainBlock represents a DAG chain-block
+type ChainChangedChainBlock struct {
```
Comment (someone235) [inline app/appmessage/rpc_notify_chain_changed.go:45]: This is used also for getChainFromBlock, so can you rename it to ChainBlock?

**Review by someone235 — CHANGES_REQUESTED** [review]

File: app/rpc/rpchandlers/notify_block_added.go [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+import (
+	"github.com/kaspanet/kaspad/app/appmessage"
+	"github.com/kaspanet/kaspad/app/rpc/rpccontext"
+	"github.com/kaspanet/kaspad/infrastructure/network/netadapter/router"
+)
+
+// HandleNotifyBlockAdded handles the respectively named RPC command
+func HandleNotifyBlockAdded(context *rpccontext.Context, router *router.Router, _ appmessage.Message) (appmessage.Message, error) {
+	listener, err := context.NotificationManager.Listener(router)
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply app/rpc/rpchandlers/notify_block_added.go:11]: I was wrong, it doesn't actually panic.

File: infrastructure/network/netadapter/server/grpcserver/grpc_connection.go [hunk 10]
```diff
@@ -139,7 +143,7 @@ func (c *gRPCConnection) closeSend() {
 	c.streamLock.Lock()
 	defer c.streamLock.Unlock()
 
-	clientStream := c.stream.(protowire.P2P_MessageStreamClient)
+	clientStream := c.stream.(grpc.ClientStream)
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply infrastructure/network/netadapter/server/grpcserver/grpc_connection.go:146]: grpc.ClientStream is simply the base type for protowire.P2P_MessageStreamClient and protowire.RPC_MessageStreamClient.

**Comment by stasatdaglabs** [comment]
Rename SendRawTransaction to SubmitTransaction.

File: infrastructure/network/rpcclient/rpc_send_raw_transaction.go [hunk 11]
```diff
@@ -0,0 +1,30 @@
+package rpcclient
+
+import (
+	"bytes"
+	"encoding/hex"
+	"github.com/kaspanet/kaspad/app/appmessage"
+)
+
+// SendRawTransaction sends an RPC request respective to the function's name and returns the RPC server's response
+func (c *RPCClient) SendRawTransaction(msgTx *appmessage.MsgTx) (*appmessage.SendRawTransactionResponseMessage, error) {
```
Comment (stasatdaglabs) [inline infrastructure/network/rpcclient/rpc_send_raw_transaction.go:10]: Rename SendRawTransaction to SubmitTransaction.

File: app/appmessage/message.go [hunk 12]
```diff
@@ -48,10 +52,45 @@ const (
 	CmdDoneIBDBlocks
 	CmdTransactionNotFound
 	CmdReject
+
+	// rpc
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply app/appmessage/message.go:56]: Created a ticket with the list.

File: app/rpc/rpchandlers/get_mempool_entry.go [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+	"fmt"
+	"github.com/kaspanet/kaspad/app/appmessage"
+	"github.com/kaspanet/kaspad/app/rpc/rpccontext"
+	"github.com/kaspanet/kaspad/infrastructure/network/netadapter/router"
+	"github.com/kaspanet/kaspad/util/daghash"
+)
+
+// HandleGetMempoolEntry handles the respectively named RPC command
+func HandleGetMempoolEntry(context *rpccontext.Context, _ *router.Router, request appmessage.Message) (appmessage.Message, error) {
+	getMempoolEntryRequest := request.(*appmessage.GetMempoolEntryRequestMessage)
```
[the remark this answers is not in the document]
Comment (stasatdaglabs) [reply app/rpc/rpchandlers/get_mempool_entry.go:12]: Created.

**Review by someone235 — APPROVED** [review]
